src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/IntellisenseQuickInfoBuilder.cs

```diff
@@ -47,18 +55,90 @@ internal static IntellisenseQuickInfoItem BuildItem(ITrackingSpan trackingSpan,
+            var semanticSpans = new List<ClassifiedSpan>();
+
+            foreach (var span in quickInfoItem.RelatedSpans)
+            {
+                await classificationService.AddSyntacticClassificationsAsync(document, span, classifiedSpans, cancellationToken).ConfigureAwait(false);
+                await classificationService.AddSemanticClassificationsAsync(document, span, semanticSpans, cancellationToken).ConfigureAwait(false);
+            }
+
+            // replace the spans from SyntacticClassifications with spans from SemanticClassifications
+            classifiedSpans.Sort((a, b) => a.TextSpan.Start.CompareTo(b.TextSpan.Start));
+            semanticSpans.Sort((a, b) => a.TextSpan.Start.CompareTo(b.TextSpan.Start));
+
+            var i = 0;
+            foreach (var sSpan in semanticSpans)
+            {
+                for (; i < classifiedSpans.Count; i++)
+                {
+                    if (classifiedSpans[i].TextSpan.Start == sSpan.TextSpan.Start)
+                    {
+                        classifiedSpans.RemoveAt(i);
+                        classifiedSpans.Insert(i, sSpan);
+                        i++;
+                        break;
+                    }
+                }
+            }
+
+            // Convert spans to textruns
+            var runs = new List<ClassifiedTextRun>();
+            var lastSpanEnd = -1;
+            foreach (var span in classifiedSpans)
+            {
+                // Add whitespace
+                if (lastSpanEnd > 0 && span.TextSpan.Start > lastSpanEnd)
+                {
+                    // find the gap between last span and current span, read the gap text from document
+                    var spanGap = new Span(lastSpanEnd, span.TextSpan.Start - lastSpanEnd);
+                    var spanText = spanGap.GetSpanText(snapshot);
+
+                    // special handling for new line, remove all the whitespace at the begining of the new line.
+                    if (spanText.StartsWith("\r\n"))
+                    {
+                        spanText = "\r\n";
+                    }
+
+                    runs.Add(new ClassifiedTextRun(ClassificationTypeNames.WhiteSpace, spanText));
+                }
+
+                lastSpanEnd = span.TextSpan.End;
+                runs.Add(new ClassifiedTextRun(span.ClassificationType, span.TextSpan.ToSpan().GetSpanText(snapshot)));
+            }
+
+            return new ClassifiedTextElement(runs);
+        }
+
         private static ClassifiedTextElement BuildClassifiedTextElement(QuickInfoSection section)
         {
             return new ClassifiedTextElement(section.TaggedParts.Select(
                     part => new ClassifiedTextRun(part.Tag.ToClassificationTypeName(), part.Text)));
         }
 
+        private static string GetSpanText(this Span span, ITextSnapshot snapshot)
+        {
+            var tSpan = snapshot.Version.CreateTrackingSpan(span, SpanTrackingMode.EdgeExclusive);
```

Comment:
Why are we ending up with spans that don't match the snapshot? This seems risky and asking for other bugs.